### ISSUE TITLE
How to register a new payment method

### ISSUE THREAD

**Issue by webdados** [woocommerce/woocommerce-gutenberg-products-block#2858]
I'm the author of an offline payment gateway plugin for WooCommerce and I want to add support for WooCommerce checkout block.

I've created the JS files and my payment method class that extends AbstractPaymentMethodType and now I'm stuck on how to initialize it / register it.

What am I missing here?

**Comment by nerrad** [comment]
Hi there!

That's great that you are creating an integration for the Checkout Block! You can find more [documentation on building the integration here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/docs/extensibility/payment-method-integration.md).

Please do take note of this though:

> Note: This API is fairly stable, but we're still really early in the checkout block release plan so it is possible this might slightly change as more payment methods are integrated and we discover areas needing improvement. So monitoring this API will be needed.

The checkout blocks are currently released as a preview, so it's important that anyone building early integrations with it are watching for changes in apis between releases of this repository (we will highlight any significant changes in changelogs).

Also to aid with your integration, you can look at the current [in use examples here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/main/assets/js/payment-method-extensions/payment-methods).

**Comment by nerrad** [comment]
I'm going to close this issue because it is more of a support request vs something needing addressed in this repository, but I'm happy to continue the conversation about your integration efforts in this issue (I'll watch for updates).

**Comment by webdados** [comment]
Hi @nerrad 

I know the APIs will probably change. I just want to build experimental support in this early stage.

I already saw the documentation (that's how I made my integration), and here https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/docs/extensibility/payment-method-integration.md#server-side-integration it says to create the class extending AbstractPaymentMethodType which I did already. I also got the JS files created.

I might be missing something, but I'm not seeing how I "register" my class against the WooCommerce Blocks plugin.
I'm including the file via PHP on my plugin and then what? How do I "announce it"?

For instance, to add my gateway to WooCommerce I do something like this:
`require_once('my_payment_gateway_class_file.php');
add_filter( 'woocommerce_payment_gateways', function( $gateways ) {
  $gateways[] = 'my_payment_gateway_class';
  return $gateways;
});`

**Comment by nerrad** [comment]
Ahh, I see. This could be an area where we improve the docs.

You'll need to register your payment method class with the payment method registry. So doing something like this:

```php
<?php
use Automattic\WooCommerce\Blocks\Payments\PaymentMethodRegistry

// PaymentMethodInstance would be an instance of your class extending AbstractPaymentMethodType.
function register_my_payment_method( PaymentMethodRegistry $payment_method_registry ) {
  $payment_method_registry->registry( PaymentMethodInstance );
}
add_action( 'woocommerce_blocks_payment_method_type_registration', 'register_my_payment_method' );
```

You'd have to add the action before the WordPress `init` action hook at priority 5 fires in the WordPress loading process.

Let me know how that goes for you and I'll update the docs after hearing back.

**Comment by webdados** [comment]
`<?php
$payment_method_instance = new MyPaymentMethod;
`

Here I'm supposed to use my new class name, the one that extends AbstractPaymentMethodType, right?

I'm getting:
`Fatal error: Uncaught ArgumentCountError: Too few arguments to function Automattic\WooCommerce\Blocks\Payments\Integrations\MultibancoIfthenPay::__construct(), 0 passed in (...) and exactly 1 expected in (...)`

**Comment by nerrad** [comment]
> Here I'm supposed to use my new class name, the one that extends AbstractPaymentMethodType, right?

Correct. 

The error means that your class constructor expects an argument on instantiating and you didn't pass anything in. So you'll need to construct it with whatever you've defined as your class needing. Please note that if you're going by the example of payment methods in this plugin, they are receiving an instance of `Automattic\WooCommerce\Blocks\Assets\Api` but you can't use that in your context because it's specific to the plugin and not something for general use. All it is, is just a helper we pass around for path discovery and asset registration in the context of the plugin. For your use case you can likely just use the `wp_register_script` api.

**Comment by webdados** [comment]
Hi, and sorry to bother again.

I've based my integration on the "Cheque" one because my payment method is actually an offline payment method.

For your reference, it works like this: 
- On "process_payment" we generate a unique payment reference and set the order "on hold"
- On the "thank you" page and on "on-hold" emails we show that reference
- The customer uses the Portuguese ATM network (via an actual ATM, home banking website, or mobile app) and pays for that reference on a specific screen that all Portuguese banks provide for "reference payment"
- The gateway calls a WC-API endpoint we created and "informs" that the payment is made
- We set the order as paid

We have 3000+ shops using the plugin at the moment: https://pt.wordpress.org/plugins/multibanco-ifthen-software-gateway-for-woocommerce/

So this is like a bank transfer on steroids with Instant Payment Notification, but still must be considered a "offline" payment method.

The plugin actually implements three Portuguese payment methods:
- Multibanco (the one that I explained above, and the only one I'm trying to integrate at the moment)
- MB WAY (we make a payment request to the gateway and the customer gets a push notification and authorizes the payment within 5 minutes on a specific banking mobile app)
- Payshop (similar to Multibanco, but the customer makes the reference payment at kiosks and similar shops, in cash)

Back to the integration:

I've now successfully registered the payment method (still using the cheque icon thou, as I wasn't able to find out how to import my own icon):

![image](https://user-images.githubusercontent.com/6853605/87531946-590c8280-c68a-11ea-86af-bc3a5d88865b.png)

When I place the order, the order is correctly set to my payment method but:
- It's created as "draft" instead of "on hold"
- No reference is generated

So it seems to me that the "process_payment" method of my payment gateway is not being called at all.

Here are my files:
- PHP Class: https://gist.github.com/webdados/4f06f5f895f246d07d387e36a63889b2
- JS main file (unminified from the cheque payment method and then adapted): https://gist.github.com/webdados/e60c4d1a4f5fa992abdf4838c55b4aba
- I also have the index.js and constants.js files, but are they used at all? I don't see them being called on the network tab on developer tools... What are they for?

Can you guys help understand what's missing here?

I know it's not your job to "teach me" the integration, but I'm convinced my doubts will help you simplify the process of adding custom payment gateways by third-party developers (and it does seem very complicated at the time, I just wish we could "declare" WooCommerce Blocks compatibility and declare some settings and all the javascript work was done by your plugin).

**Comment by nerrad** [comment]
I am willing to help but I just want to be clear that there may be some delays in responses because I have a number of things on my plate competing for attention - so I hope you'll understand if there's a potential 2-3 day delay in future responses.

Thanks for sharing the code you shared, but unfortunately that doesn't give enough to really effectively troubleshoot with you.

- The php class you shared looks fine for registering your payment method with the blocks server side (primarily around loading the new assets).
- The JS file, even though it's unminified is really hard to follow. I don't recommend setting up your payment method by simply adapting an existing payment method file. The `index.js` and `constants.js` file are the _source_ files for the payment method which are built into the final distributed file (the file you are adapting from) using a build process (which uses Webpack). The payment methods registration and components use React - so you will need some familiarity to register a payment method via React.
- I suspect the `process_payment` method for your payment gateway is not getting called because the `paymentMethodId` property (which defaults to `name` in the client side payment method registration) doesn't match the expected slug for Woo to know what Woo Gateway class you've created to call.

 > and it does seem very complicated at the time, I just wish we could "declare" WooCommerce Blocks compatibility and declare some settings and all the javascript work was done by your plugin

I realize for those unfamiliar with modern JavaScript (ESNext) and React that it can appear complicated and difficult - we're not intentionally trying to make things more difficult and in reality, by using React and components we are making the payment method registration more streamlined and modular for client side registration. 

Practically, its near impossible to just have "all the javascript work done by the plugin" for _every_ payment method out there.  There's simply no way we could account for all the variations in how payment methods process payments or surface to the users to interact with on the frontend.

**Comment by webdados** [comment]
> I am willing to help but I just want to be clear that there may be some delays in responses because I have a number of things on my plate competing for attention - so I hope you'll understand if there's a potential 2-3 day delay in future responses.

No problem. This is a test for now.

> I suspect the process_payment method for your payment gateway is not getting called because the paymentMethodId property (which defaults to name in the client side payment method registration) doesn't match the expected slug for Woo to know what Woo Gateway class you've created to call.

That's odd because the order was indeed registered in my payment method:
![image](https://user-images.githubusercontent.com/6853605/87547565-3175e400-c6a3-11ea-970e-c838389f7bc1.png)
Only order 16389 was completed with the checkout blocks.

When submitting the checkout the correct value is passed on `$_POST['payment_method']`:
![image](https://user-images.githubusercontent.com/6853605/87547912-b2cd7680-c6a3-11ea-9c39-625563fc2e2f.png)

**Comment by webdados** [comment]
I found the problem why my `process_payment` method is not called.

On Checkout.php you only run your `process_payment` method (which will then call the original payment method) if the order `needs_payment` (line 170).

An order is only considered as needing payment if the status is "pending", "failed", and (added by your plugin) "checkout-draft" (which is the case).

The problem is that we also hook into the `woocommerce_valid_order_statuses_for_payment` filter to add "on-hold" if the order has our payment method. In doing so (after your hook), we are incorrectly ditching the "checkout-draft" item from the array.

We do this (add "on-hold") because although our payment method is "offline", it does need payment.

So we added "checkout-draft" to our valid needs payment statuses and... IT WORKS!

Next question: how do I set a custom icon? :-D

**Comment by nerrad** [comment]
> So we added "checkout-draft" to our valid needs payment statuses and... IT WORKS!

🎉 

> Next question: how do I set a custom icon?

The `label` property can be any React node (instead of a text string).

**Comment by webdados** [comment]
:-)

![image](https://user-images.githubusercontent.com/6853605/87675903-dd850100-c76f-11ea-83da-54aaf3bbaa18.png)

It's working.

Now I need to implement some rules I already have in place on the "regular checkout", to make the payment available or not, like:
- Minimum order value
- Maximum order value
- Country of the customer

Would you recommend using the "is_active" PHP class method somehow, or client-side via the Props Fed to Payment Method Nodes?

**Comment by nerrad** [comment]
> It's working.

🎉 

The `canMakePayment` property of the configuration object on the client side payment registration is what you'll want to use here. As noted [from the docs](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/docs/extensibility/payment-method-integration.md#express-payment-methods---registerexpresspaymentmethod-paymentmethodcreator-) this is:

> A callback to determine whether the payment method should be available as an option for the shopper. The function will be passed an object containing data about the current order. Return a boolean value - true if payment method is available for use. If your gateway needs to perform async initialisation to determine availability, you can return a promise (resolving to boolean). This allows a payment method to be hidden based on the cart, e.g. if the cart has physical/shippable products (example: Cash on delivery); or for payment methods to control whether they are available depending on other conditions. Keep in mind this function could be invoked multiple times in the lifecycle of the checkout and thus any expensive logic in the callback provided on this property should be memoized.

We don't currently document what the callback receives as arguments because that might change as we iterate, but you can see what it is currently fed here:

https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/37b757ec9b2141658ec5c667f6b47609fad0cece/assets/js/base/context/cart-checkout/payment-methods/use-payment-method-registration.js#L68-L73

The return value from `canMakePayment` determines whether the payment method appears as an option in the checkout block. Currently, the checkout process calls registered payment methods `canMakePayment` callback when it checkout is initializing, and also anytime `selectedShippingMethods` change.

Thus you should be able to get the cartTotals and use that to match against the minimum order value and maximum order value options you provide from your payment method.

Country of the customer should be something you can derive from the `shippingAddress` as well.

**Comment by webdados** [comment]
👍
Back to the code.

**Comment by webdados** [comment]
Update.

I've managed to return true/false on `canMakePayment` depending on the cart value and the fact the currency is euro or not (useful for multi-currency websites).

I can also do it based on the shipping address country, but I only have access to the shipping address and not the billing address.
Let's see this scenario: a person from Portugal is buying a gift for a friend in the US. The shipping address will be in the US but he will deactivate the "Use same address for billing" option and enter his Portuguese address on the billing address. I should allow him to use a Portuguese payment gateway, don't you agree?

Another problem is that `canMakePayment` is not actually being called when the addresses change, or the shipping method is changed (@nerrad told me it should, maybe there's a bug?) or even when a coupon is applied and the cart value changes because of that, so all the tests the payment gateway is making on `canMakePayment` are firing only when the checkout block loads and not when there are changes that can affect those changes.

Just like the regular checkout `update_checkout` event, this should run everytime anything changes on the checkout:

- Activate/deactivate billing address
- Change any address field (country, state, postcode, city, ...)
- The cart value changes because of coupons or shipping method or anything else
- The shipping method changes (this should be working already based on @nerrad comment but it's not, at least on my side)

**Comment by webdados** [comment]
https://woocommerce.wordpress.com/2020/07/22/woocommerce-blocks-3-0-release-notes/

> Prior to 3.0, the callback provided on this property by the payment method would only ever get invoked once during a checkout session. However, with 3.0 this could now get invoked multiple times during the checkout session. This is so payment methods have the opportunity to update whether they can show in the ui/ux or not depending on changes in the checkout itself (shipping rates selected, totals etc).

❤️

I think we still need to have access to the "billing address details".
Should I open a new issue requesting this?

**Comment by nerrad** [comment]
> I think we still need to have access to the "billing address details".
> Should I open a new issue requesting this?

Yes please do, if you could also include in the issue details around the use-cases for why this additional data is useful, that will help ensure we account for that properly if/when we add it.
